development/webpack/README.md

```diff
@@ -0,0 +1,179 @@
+
+- using JavaScript (this tool is only fast because it uses [SWC](https://swc.rs/) for compilation, which is written in
+  Rust)
+- requiring/importing large libraries
+- functional programming paradigms (JavaScript is not Haskell after all)
+  - like chaining map, filter, reduce, etc. when a single loop would do.
+  - try to avoid looping over the same data/file multiple times
+- using async IO when sync IO would do
+  - non-blocking IO is great, but not when it's the only IO happening at the time and we don't care about blocking the
+    main process.
+- launching shells, workers, or other processes without measuring the cost
+- unnecessary IO
+- validation, linting, or other checks that are not necessary
+
+If you must add something that slows it down, consider putting it behind a flag. If it must be in the default mode, try
+to run it in parallel with other tasks.
+
+### The Cache
+
+The build process uses a cache to speed up successive builds. The cache is stored in the `node_modules/.cache/webpack`
+directory.
+
+The cache is slow. Very slow. It takes about 50% of the total time just to create the cache. But you shouldn't notice
+that because the caching step is pushed to a background process.
+
+The way this works is by running the build in a background child process, and then detaching that child process from the
+parent process once the build is complete (and cache reconciliation and persistance begins).
+
+Launching the build in a background process does take time, but its much less time than cache creation, so it works out.
+
+The child process is run with its own TTY for `stderr` and `stdout`; the child's stdio dimensions are kept in sync with
+the parent's, and all TTY features of the parent are available in the child (formatting, colors, cursors, etc.). On
+Windows an IPC channel is used to communicate between the parent and child processes, on \*nix this is done via signals.
+The parent process listens for the child process and signal the parent, and when it does, the parent disconnects from
+the child and shuts down, leaving the child to run in the background so the cache can be processed and persisted.
+
+### To do:
+
+- define and wrangle the difference between `lockdown` and `lavamoat` options.
+- MV3 support.
+  - Service workers, used by MV3, must load dependencies via `importScripts`.
+  - there are existing webpack plugins that do this, but they are not yet integrated into this build tool and would
+    require changes to our code and existing gulp-based build process to work.
+- Make lavamoat work so we can run production builds
+- Make LiveReload, Hot Module Reloading, and/or React Refresh work
+  - prerequisite: https://github.com/MetaMask/metamask-extension/issues/22450
+- Make the build tool even faster (switch to RSPack once it hits 1.0.0?)
+- enable `yarn webpack completion`. It doesn't work with multiple-word commands (`yarn webpack ...`) and is currently disabled.
```

Comment:
This doesn't currently work because of a bug in yargs. I think the yargs maintainer will accept a PR to fix it, which I plan on doing Soon ™️ .